src/openlifu/plan/solution.py

```diff
@@ -264,6 +232,47 @@ def analyze(self, transducer: Transducer, options: SolutionOptions = SolutionOpt
 
         return solution_analysis
 
+    def scale(
+            self,
+            transducer: Transducer,
+            focal_pattern: FocalPattern,
+            analysis_options: SolutionAnalysisOptions = SolutionAnalysisOptions()
+    ) -> SolutionAnalysis:
```

Comment:
done

Reply:
The test should verify that the scaling function works correctly and does the correct thing, not just that it runs. But I see that this might be difficult because you have to create a solution whose analysis looks juuust right. In this case, it makes sense to split up the `scale` function into testable components. We should have a unit test that specifically confirms that the pulse scaling code

```py
        scaling_factors = np.zeros(self.num_foci())
        for i in range(self.num_foci()):
            focal_pattern_pressure_in_MPa = focal_pattern.target_pressure * getunitconversion(focal_pattern.units, "MPa")
            scaling_factors[i] = focal_pattern_pressure_in_MPa / analysis.mainlobe_pnp_MPa[i]
        max_scaling = np.max(scaling_factors)
        v0 = self.pulse.amplitude
        v1 = v0 * max_scaling
        apod_factors = scaling_factors / max_scaling

        for i in range(self.num_foci()):
            scaling = v1/v0*apod_factors[i]
            self.simulation_result['p_min'][i].data *= scaling
            self.simulation_result['p_max'][i].data *= scaling
            self.simulation_result['ita'][i].data *= scaling**2
            self.apodizations[i] = self.apodizations[i]*apod_factors[i]
        self.pulse.amplitude = v1
```

does what it's supposed to do. If we need to split that code out into a separate function in order to unit test it, then that's better I think

Reply:
hmm I see, the issue is I am not sure how to create the test data for those use case that is why at the moment I just tested the run as for solution.analysis

Reply:
For this section of code, the inputs are
- a FocalPattern
- the number of foci
- a list of `mainlobe_pnp_MPa` values
- a pulse amplitude

and we can say the output is
- a list of `scaling` values
- a list of of `apod_factors`
- a new pulse amplitude `v1`

so this portion of `Solution.scale` could be refactored to something like

```
        scalings, apod_factors, v1 = a_new_standalone_function_that_can_be_unit_tested(focal_pattern, num_foci, mainlobe_pnp_MPa_list, pulse_amplitude)
        for i in range(self.num_foci()):
            self.simulation_result['p_min'][i].data *= scalings[i]
            self.simulation_result['p_max'][i].data *= scalings[i]
            self.simulation_result['ita'][i].data *= scalings[i]**2
            self.apodizations[i] = self.apodizations[i]*apod_factors[i]
        self.pulse.amplitude = v1
```

(perhaps with better variable names :slightly_smiling_face:)

Reply:
For this please just refactor a bit as described in the conversation, so that we have a function that can be unit tested. But you don't need to add the unit test.

Reply:
Done